### PR TITLE
Backend support (in flush) for newline : false

### PR DESCRIPTION
**Pull request by hoodmane**
The current behavior of `term.echo(str, {newline : false})` is pretty buggy. For example: `term.error(str, {newline : false})` doesn't produce red text, it interferes with `term.get_prompt` which will show the partially echoed line of text and `term.set_prompt` which won't work correctly at all, it seems to be incompatible with a bunch of the other options (I don't expect finalize to work). The `ENTER` handler it provides also causes weird effect.

The problem is that the text echoed with `newline : false` is handled in a hacked-on extension. In order to get the behavior to be consistent and to make `newline : false` interact correctly with other formatting, the feature should be implemented in the core rendering engine and not as an add-on.

This PR moves the implementation of the `newline` option into the main rendering engine. The main idea is to output the partial line as normal but whenever the output ends in a partial line of text, update the style for the command line to move it up and the first line of it over to position the input field to the right of the final partial output field.

### CODE REVIEW

File: css/jquery.terminal.css [hunk 1]
```diff
@@ -684,14 +684,18 @@ terminal .terminal-output > div {
     margin-bottom: 10px;
     position: relative;
 }
+div[data-index] > div {
+    width: -moz-fit-content;
+    width: -webkit-fit-content;
+    width: fit-content;
+}
```
Comment (jcubic) [inline css/jquery.terminal.css:691]: It will not work in IE, but I think that this not a big deal. This only need to be documented. that echo newline don't work with IE.

**Review by jcubic — COMMENTED** [review]
Thanks for the PR, I will need to take a look little big closer how this work. And how it will behave with line wrapping, probably monkey patch in echo_newline.js also didn't handled that properly. The coverage dropped, maybe it would be good idea to also write unit tests for flush, that is missing coverage.

File: js/jquery.terminal-src.js [hunk 2]
```diff
@@ -9318,21 +9344,56 @@
...
+                                var div = document.createElement('div');
+                                div.innerHTML = line;
+                                // width = "100%" does some weird extra magic
+                                // that makes the height correct. Any other
+                                // value doesn't work.
+                                div.style.width = "100%";
+                                if (data.newline) {
+                                    div.className = 'cmd-end-line';
+                                }
+                                wrapper[0].appendChild(div);
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply js/jquery.terminal-src.js:9365]: > I have no idea why

There is a good reason for this: I am not very familiar with jQuery. Will fix!

[the remark this answers is not in the document]
Comment (hoodmane) [reply js/jquery.terminal-src.js]: Ok I think I fixed this.

**Comment by hoodmane** [comment]
>  line wrapping, probably monkey patch in echo_newline.js also didn't handled that properly

Surprisingly, the old method handles line wrap just fine. Apparently it's okay to set really long prompts:

image

Line wrap is handled fine with this approach too:

image

**Comment by hoodmane** [comment]
> The coverage dropped, maybe it would be good idea to also write unit tests for flush, that is missing coverage.

Actually flush isn't missing coverage: most of the lines are covered 1000+ times and the only part missing coverage is the catch clause. Looking at the coverage report, every line that I added is covered a large number of times -- I was adjusting very hot functions. I don't really understand why the coverage went down.

File: css/jquery.terminal.css [hunk 3]
```diff
@@ -684,14 +684,18 @@ terminal .terminal-output > div {
...
     .terminal,
     .terminal-output > :not(.raw) span[data-text]:not(.token):not(.inverted):not(.terminal-inverted):not(.cmd-inverted):not(.terminal-error):not(.emoji),
     .terminal-output > :not(.raw) a,
-    .terminal-output > :not(.raw) div,
-    .cmd,
-    .cmd span[data-text]:not(.cmd-inverted):not(.token):not(.emoji),
-    .cmd div {
+    .terminal-output > :not(.raw) div
+    {
         color: var(--color, #aaa);
```
Comment (jcubic) [inline css/jquery.terminal.css:699]: Why this change?

**Comment by jcubic** [comment]
Few more issues I've found.
* Removing background from prompt is not good idea, because it need to work on it's own outside of terminal.
This is better approach:
```css
.terminal .partial {
    position: relative;
    z-index: 400;
}
.terminal div.partial > div {
    width: -moz-fit-content !important;
    width: -webkit-fit-content  !important;
    width: fit-content  !important;
}
```

While you modify CSS please modify `-src.css` file and always put .terminal prefix before the selectors.

* Please update the JS file and replace mixed double and single quotes with just single quotes. I should write about this in contribution guide I didn't use ESLint rules because sometimes is better ot use double quotes (e.g. when you write error message and put single quote inside like `don't` word).
* If you can please use snake case for variables (e.g. `lastRow` should be `last_row`).

I don't have any other issues, good job, thanks for the PR

PS: I didn't added the code into the library itself because it needed to be monkey patch, but I didn't know about `content-fit` value.

**Comment by hoodmane** [comment]
> Removing background from prompt is not good idea...  better approach:
> ```z-index: 400```

That is better!

> I didn't know about content-fit value.

I'm really confused about why this is necessary. I don't understand the behavior of the measured height and width of elements very well at all, I always just get things to work by trial and error. In particular, do you know why `width='100%'` changes the height of the empty lines? I understand that you are putting a `before::` with a zero with space, but why is `width='100%'` needed? Weirder still, it doesn't work if it's `99%` or `101%` and it only works as an inline style. Also, why isn't the div sized to fit the content by default?!
/rant

**Comment by hoodmane** [comment]
Wait:
```css
.terminal .partial {
    position: relative;
    z-index: 400;
}
.terminal div.partial > div {
    width: fit-content  !important;
}
```
Does this screw up the height of empty lines? Also, does `!important` override inline styles? I don't think this will work.

**Comment by hoodmane** [comment]
Oh I get it, you only apply the `fit-content` to the `.partial` div and I don't think the `.partial` one needs `width : 100%`. Only thing is to double check that `term.echo("\n",{newline : false})` works but that should be okay.

**Comment by hoodmane** [comment]
Yeah this is broken when `newline : false` is used with a trailing newline. But I can just detect that, remove the trailing newline, and set `newline` to true.

**Comment by hoodmane** [comment]
@jcubic Okay I think I addressed your comments?

I tried to ensure that exactly one output div appears with each 'data-index' value. Currently consecutive calls to `echo` with `newline : false` get batched together with one call with `newline : true`. This makes the behavior of the `update` function a bit unpredictable (how many lines of text will it replace?), but `term.update` was already a bit odd.

**Comment by jcubic** [comment]
I'm thinking about all those inline styles you're adding, there is issue to make jQuery Terminal work with strict CSP maily inline styles. I'm planing on removing the inline styles at all. I'm wondering if it will be possible to change your code into `<style></style>` even with current code it will be difficult, adding more inline style don't make things simpler.

Maybe there is better way then this, it seems that you just move the cmd to put above last line (didn't notice that first time when check the code), I think this is worse then modifying the prompt. Also I don't like exposing hidden `__set_prompt_margin` function.

This is why I didn't wanted to put that code into the terminal because your solution is also one big hack, like old `echo_newline.js`. I didn't wanted to put such hacks into the code.

Usually when I want echo without the newline I don't use `echo_newline.js` I use buffer defined in my own code outside of jQuery Terminal. Usually I don't want to left the code before the cursor like in Bash I want to add implicit newline when.
You can see how this work in my lisp interpreter https://lips.js.org/beta.html
Type
```scheme
(display "x") (display "y") (newline)
```
and 
```scheme
(display "x") (display "y")
```
both work the same because there is implicit newline after the flush (flush in my code not in jQuery Terminal).

I think this is better then broken command line like in bash when someone don't echo newline at the end.
Here is the code for my terminal app https://github.com/jcubic/lips/blob/master/lib/js/terminal.js

**Comment by hoodmane** [comment]
Does that approach work with sleep? I'm concerned about something like
```(display "x") (sleep 10) (display "y")``` 
if you buffer the output then the first `x` won't be displayed while sleeping. A lot of existing Python programs with interactive code (say using `input`) rely on the normal behavior of the console in order to work correctly, in particular they rely on partial lines being displayed. `input("Enter your name: ")` in python does effectively:
```python
print("Enter your name: ", end="")
sys.stdin.readline()
```
and in order to make code like this behave correctly it's helpful for the terminal to behave similar to a native terminal.

**Comment by hoodmane** [comment]
> I'm wondering if it will be possible to change your code into <style></style> even with current code it will be difficult, adding more inline style don't make things simpler.

I think I can update it to remove the inline styles, except for the `width : 100%` hack which was in the original code and I have no replacement for it. If this could work without inline styles, would you consider including it? I think this is a huge improvement over hacking the prompt because:
1. it supports normal styling and options (with the prompt hack, newline : false breaks the flush option, the finalize option, and others, term.error("some error", {newline : false}) is not displayed in red).
2. it breaks get_prompt and set_prompt
3. it interacts badly with term.push, term.pop, term.read, term.login... basically everything

**Comment by jcubic** [comment]
I will think about this, you did lot of great work. I will test how this works with my Scheme REPL in Scheme that give me problems with old echo_newline.js

**Comment by jcubic** [comment]
Hi, I decided that I will merge your code. I will need to check why the drop of 0.07 in coverage thoush. It's getting close to 82 after rounding.

I will yet again documentation issue to document that `newline: false` feature is experimental.

**Comment by hoodmane** [comment]
Would you like me to try to modify to remove the added inline styles?

**Comment by jcubic** [comment]
You can try but I don't think it's really possible with current approach. You need dynamic top and left to move the cmd.

I've found one issue with the your solution. If you echo without newline you can't select text of the prompt. The problem is that div with partial class is on top of the cmd.

This is the fix:
```css
.terminal div.partial > div, .terminal div.partial {
    width: 100% !important;
    width: -moz-fit-content !important;
    width: -webkit-fit-content !important;
    width: fit-content !important;
}
```

**Comment by jcubic** [comment]
I will also need to check how this will work on mobile phone.

**Comment by jcubic** [comment]
Another bug is in copy/paste terminal is hijacking the right click over hidden textarea to allow pasting of text. This don't work anymore if you echo without newline.

**Comment by hoodmane** [comment]
> You can try but I don't think it's really possible with current approach

Well if you had a css variable `--line-height` you could do 
```css
.cmd .shift_up {
    top : calc(-var(--line-height));
}
```
though this would screw up if the partial line contained an unusually tall unicode character. For the width you could add an extra `<span>` to the left of the command prompt with:
```css
.cmd .prompt-margin {
    display: inline-block;
    user-select : none;
    visibility : hidden;
}
```
and set the text in the span like `promptMarginSpan.innerText = partialLineInnerText`.

**Comment by hoodmane** [comment]
I think there is also a bug if the prompt is empty.

**Comment by hoodmane** [comment]
Any update on this?

**Comment by jcubic** [comment]
Sorry didn't have time to test locally how this works. Will merge into experimental branch, since this is not 100% ready.

**Comment by jcubic** [comment]
Will check how this work and maybe fix some issues, maybe tomorrow will find some time.

**Comment by hoodmane** [comment]
Thanks!

**Comment by jcubic** [comment]
I was testing your solution locally, at it seems that it don't work at all, it don't move the cmd to the left so text that is echo without newline is not visible.

**Comment by jcubic** [comment]
It works great, thanks for working on this. I've added minor change to get rid of 0.01px border, I've added zero space width prompt when it's empty.

**Comment by jcubic** [comment]
About copy/paste it's also broken on devel branch.

**Comment by jcubic** [comment]
New version 2.23.0 with the feature was just released.

**Comment by hoodmane** [comment]
Thanks! I'm really glad this worked out.

**Comment by jcubic** [comment]
Add one fix to your hack to force reflow in Firefox, it was not shoring the cursor in one of my projects, I've set the timeout to 0 but move the showing timeout into the first timeout.

PS: what tools did you use to show the coverage? I've missed that first time when you added that screenshot.
